feat: add panel support for BrowserWindow

isailaandrei opened the pull request:
#### Description of Change
Created a new ElectronNSPanel window which inherits from ElectronNSWindow. This window mimics the behavior of NSPanel - visible on top of full screened apps, joins all spaces. To achieve this it adds the "NSWindowStyleMaskNonactivatingPanel" style mask, normally reserved for NSPanel, at runtime. 

This is a workaround to avoid patching chromium and duplicating lots of window code.

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [ ] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

notes: Added support for panel-like behavior. Window can float over full-screened apps.

codebytere reviewed (changes requested):
Hi! Thanks for opening this PR - however, all new PRs, especially semver-minor or major PRs, must target `main`. Only then, when it's reviewed and merged, can it be backported to version branches.

Could you also talk a bit more in your PR body about the impetus for this PR? What use cases are you trying to address?

mitchchn commented:
Resolves #31538, #4939.

I really like this implementation of panels overall, as it's similar to third-party solutions but safer and cleaner.

**Pros**

- Eliminates the need for a third-party native module
- Avoids runtime swizzling
- Does not require patching Chromium + duplicating a ton of window code
- API would be forward compatible if the implementation ever changed to use a true NSPanel

**Cons**

- This approach to emulating NSPanel may not work in future versions of macOS 

Non-activating panels are such a useful and desirable part of modern apps (floating palettes are everywhere!) and I think the risk is manageable.

Still, it would be great to work toward replacing this implementation with a real NSPanel in the long term. If done right it would be a transparent change to users who adopt this API in its current form.

isailaandrei commented:
Hey @codebytere

I changed the target branch to "main" as suggested and changed the wording in browser-window to make it more clear what this new type achieves i.e makes the window float on top of full-screened apps.

As per @mitchchn's comment, this pull request addresses a feature which has been previously requested. This is still obviously not the ideal solution, but a workaround.

isailaandrei commented:
Hey all,

We started using this patch internally and we found a weird bug. If we toggle the "resizable" property on BrowserWindow with type = "panel" the height of the window increases by 28 points (which is height of titlebar on Mac). 

I confirmed in Xcode that this is not an a bug on the native implementation, so something is going wrong when SetCanResize and then Widget->OnSizeConstraintChanged() is called in chromium. I'm still looking into this, but wanted to share here in case anyone has experience with this and can pinpoint the issue more easily. 



Fiddle code to repro:

  const mainWindow = new BrowserWindow({
    width: 200,
    height: 200,
    frame: false,
    type: "panel"
  });
  
  let resizable = true;
  setInterval(() => {
    resizable = !resizable;
    mainWindow.setResizable(resizable);
    console.log(mainWindow.getBounds());
  }, 2000);

codebytere reviewed (approved):
API LGTM

miniak commented:
/trop run backport-to 20-x-y

Hupka commented:
Hey folks! 

A question: this implementation indeed makes the window/panel appear without taking focus. But it now also can't react to hover events anymore... 🤔 Would you know which additional properties would do this? 

Best,
Adrian

OmarHedeya95 commented:
NSWindow does not support nonactivating panel styleMask 0x80

acemtp commented:
Hello,

Is it normal that the official doc has no information about type: "panel" ?

I spent lot of time figuring how to do that :-)

jbriales commented:
> Hello,
> 
> Is it normal that the official doc has no information about type: "panel" ?
> 
> I spent lot of time figuring how to do that :-)

Agree. It seems it's documented in https://www.electronjs.org/docs/latest/api/structures/base-window-options and even then it's hard to find, since one wouldn't know up-front this kind of window is a "panel".

Maybe adding a corresponding section within https://www.electronjs.org/docs/latest/tutorial/window-customization would help with discoverability here?
If that's deemed valuable I can try and add it myself :)

jbriales commented:
Also @isailaandrei what's the expected use of this `type="panel"` option wrt others like `frame: false` or `resizable: false`.
Are they incompatible/overriden/ignored?

jbriales commented:
On close inspection this does not seem to be working for me (I was confused because I still had other options that made the window look like a panel ON).

I'm seeing this when running a toy example:
```
Electron[56400:9037088] NSWindow does not support nonactivating panel styleMask 0x80
```

Any hints? I'm on Sonoma 14.5

acemtp commented:
We also have this message and it works fine, it's just informative.

jbriales commented:
After some days using this, it seems it generally works although still it has some rough edges.
E.g. (it seems) by using a normal window vs actual panel under the hood it doesn't get transparent to tools using the Accessibility API such as the Yabai WM: https://github.com/koekeishiya/yabai/issues/2397
This causes some edge-cases and problems in there.

@isailaandrei are there any ideas in the horizon from initial discussions on how to achieve a more panel-like identity...?
